File: Farmsy/Features/Trips/ShoppingListSheet.swift
import SwiftUI
import CoreLocation

/// "I need eggs, milk and potatoes — where do I drive?"
///
/// The map answers where a farm is. This answers which farms to visit for a
/// list, and in what order. The result is ordinary trip stops, so the route
/// line, the totals, the corridor list and the Google Maps hand-off all pick it
/// up unchanged.
///
/// A picker rather than a text field. The words that mean a product are served
/// (GET /api/shopping/items) because guessing them locally is what sends
/// somebody asking for lamb to a beef farm; picking from the list means the
/// terms are the same ones the website matches on.
///
/// The matching itself runs on pins already in memory, so an empty result means
/// "no farm near you says it sells this", not "we could not look".
struct ShoppingListSheet: View {
    @Environment(\.dismiss) private var dismiss
    @Environment(FarmsStore.self) private var farms
    @Environment(TripStore.self) private var trip

    /// The starting point to plan from — the trip's origin, or the device.
    let origin: CLLocationCoordinate2D

    @State private var catalogue = ShoppingItems.shared
    @State private var plan: ShoppingPlanner.Plan?
    @State private var isPlanning = false

    private var picked: [ShoppingItem] {
        trip.wantedProducts.compactMap { catalogue.item(id: $0) }
    }

    var body: some View {
        VStack(spacing: 0) {
            header
            ScrollView(showsIndicators: false) {
                VStack(alignment: .leading, spacing: 16) {
                    if catalogue.items.isEmpty {
                        loading
                    } else {
                        picker
                        if let plan { result(plan) }
                    }
                }
                .padding(.horizontal, 16)
                .padding(.top, 14)
                .padding(.bottom, 20)
            }
            actions
                .padding(.horizontal, 16)
                .padding(.bottom, 14)
        }
        .background(Color.cream.ignoresSafeArea())
        .task { await catalogue.loadIfNeeded() }
        // The produce text is what every match is made against — without it the
        // sheet would honestly report that nothing is sold anywhere.
        .task { await farms.loadFlagsIfNeeded() }
    }

    // MARK: - Pieces

    private var header: some View {
        HStack {
            VStack(alignment: .leading, spacing: 2) {
                Text("Shopping list").font(.ui(18, .bold)).foregroundStyle(Color.ink)
                Text("We pick the farms, you drive once")
                    .font(.ui(12)).foregroundStyle(Color.inkMuted)
            }
            Spacer()
            Button { Haptics.tap(); dismiss() } label: {
                Image(systemName: "xmark")
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(Color(hex: 0x6B7280))
                    .frame(width: 32, height: 32)
                    .background(Color(hex: 0xF3F4F6), in: Circle())
            }
            .buttonStyle(.plain)
        }
        .padding(.horizontal, 16)
        .padding(.top, 16)
    }

    @ViewBuilder
    private var loading: some View {
        if catalogue.loadFailed {
            Text("The product list couldn't be loaded. Check your connection and try again.")
                .font(.ui(14)).foregroundStyle(Color.inkMuted)
        } else {
            ProgressView().tint(Color.farmGreen).frame(maxWidth: .infinity).padding(.vertical, 24)
        }
    }

    private var picker: some View {
        FlowRow(spacing: 8) {
            ForEach(catalogue.items) { item in
                let on = trip.wantedProducts.contains(item.id)
                Button {
                    Haptics.tap()
                    trip.toggleProduct(item.id)
                    plan = nil          // the answer belonged to the old list
                } label: {
                    HStack(spacing: 6) {
                        if on {
                            Image(systemName: "checkmark").font(.system(size: 10, weight: .bold))
                        }
                        Text(item.label).font(.ui(14, .medium))
                    }
                    .foregroundStyle(on ? .white : Color.ink)
                    .padding(.vertical, 8).padding(.horizontal, 12)
                    .background(on ? Color.farmGreenMap : .white, in: Capsule())
                    .overlay(Capsule().stroke(on ? Color.clear : Color.hairline, lineWidth: 1))
                }
                .buttonStyle(.plain)
            }
        }
    }

    @ViewBuilder
    private func result(_ plan: ShoppingPlanner.Plan) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            if plan.isEmpty {
                Text("No farm within 25 km lists any of this.")
                    .font(.ui(14)).foregroundStyle(Color.inkMuted)
            } else {
                ForEach(Array(plan.picks.enumerated()), id: \.element.osmId) { i, pick in
                    HStack(alignment: .top, spacing: 12) {
                        Text(verbatim: "\(i + 1)")
                            .font(.ui(13, .bold)).foregroundStyle(.white)
                            .frame(width: 24, height: 24)
                            .background(Color.farmGreenMap, in: Circle())
                        VStack(alignment: .leading, spacing: 2) {
                            Text(farms.pin(forOsmId: pick.osmId)?.name ?? pick.osmId)
                                .font(.ui(15, .semibold)).foregroundStyle(Color.ink).lineLimit(1)
                            Text(labels(pick.covers).joined(separator: " · "))
                                .font(.ui(13)).foregroundStyle(Color.inkMuted).lineLimit(2)
                        }
                        Spacer(minLength: 0)
                    }
                }
            }
            // Said out loud rather than quietly dropped: a list that half worked
            // is only useful if you know which half.
            if !plan.missing.isEmpty {
                Text("Not found nearby: \(labels(plan.missing).joined(separator: ", "))")
                    .font(.ui(13, .medium)).foregroundStyle(Color.inkMuted)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .card()
    }

    private var actions: some View {
        VStack(spacing: 10) {
            Button {
                Haptics.tap()
                if let plan, !plan.isEmpty { addToTrip(plan) } else { buildPlan() }
            } label: {
                HStack(spacing: 8) {
                    if isPlanning {
                        ProgressView().tint(.white)
                    } else {
                        Image(systemName: plan?.isEmpty == false ? "plus" : "sparkles")
                            .font(.system(size: 13, weight: .semibold))
                    }
                    // Two Text views rather than a ternary inside one: each
                    // literal is then its own catalog key, which a ternary of
                    // interpolated literals is not guaranteed to be.
                    if let plan, !plan.isEmpty {
                        Text("Add \(plan.picks.count) stops to my trip").font(.ui(14, .semibold))
                    } else {
                        Text("Plan my trip").font(.ui(14, .semibold))
                    }
                }
                .foregroundStyle(.white).frame(maxWidth: .infinity).padding(.vertical, 14)
                .background(Color.farmGreenMap.opacity(canPlan ? 1 : 0.5),
                            in: RoundedRectangle(cornerRadius: 16))
            }
            .buttonStyle(.plain)
            .disabled(!canPlan)

            if !trip.wantedProducts.isEmpty {
                Button("Clear list") { Haptics.tap(); trip.clearProducts(); plan = nil }
                    .font(.ui(13, .medium)).foregroundStyle(Color.inkMuted)
                    .buttonStyle(.plain)
            }
        }
    }

    private var canPlan: Bool { !trip.wantedProducts.isEmpty && !isPlanning }

    // MARK: - Actions

    /// Ids back to the words on the chips, so the answer is read in the same
    /// language it was asked in.
    private func labels(_ ids: [String]) -> [String] {
        ids.map { catalogue.item(id: $0)?.label ?? $0 }
    }

    private func buildPlan() {
        isPlanning = true
        // The flags payload may still be arriving on a cold open; waiting for it
        // is the difference between "nothing sells eggs" and "we have not looked".
        Task {
            await farms.loadFlagsIfNeeded()
            await catalogue.loadIfNeeded()
            let candidates = farms.pins.map {
                ShoppingPlanner.Candidate(osmId: $0.osmId,
                                          coord: $0.coordinate,
                                          sells: farms.produceByOsm[$0.osmId] ?? "")
            }
            plan = ShoppingPlanner.plan(wanted: picked, farms: candidates, origin: origin)
            isPlanning = false
        }
    }

    private func addToTrip(_ plan: ShoppingPlanner.Plan) {
        trip.addStops(plan.picks.map(\.osmId))
        trip.requestFit()
        Haptics.success()
        dismiss()
    }
}

// MARK: - Flow layout

/// Chips that wrap. `Layout` rather than a stack of rows so the wrapping is the
/// real width, not a guess at how many fit.
struct FlowRow: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let maxWidth = proposal.width ?? .infinity
        var x: CGFloat = 0, y: CGFloat = 0, lineHeight: CGFloat = 0
        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if x > 0 && x + size.width > maxWidth {
                x = 0; y += lineHeight + spacing; lineHeight = 0
            }
            x += size.width + spacing
            lineHeight = max(lineHeight, size.height)
        }
        return CGSize(width: maxWidth == .infinity ? x : maxWidth, height: y + lineHeight)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        var x = bounds.minX, y = bounds.minY, lineHeight: CGFloat = 0
        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if x > bounds.minX && x + size.width > bounds.maxX {
                x = bounds.minX; y += lineHeight + spacing; lineHeight = 0
            }
            view.place(at: CGPoint(x: x, y: y), proposal: ProposedViewSize(size))
            x += size.width + spacing
            lineHeight = max(lineHeight, size.height)
        }
    }
}

File: android/app/src/main/java/app/farmsy/android/features/trips/ShoppingListSheet.kt
package app.farmsy.android.features.trips

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.farmsy.android.LocalFarms
import app.farmsy.android.LocalSession
import app.farmsy.android.LocalTrip
import app.farmsy.android.R
import app.farmsy.android.core.FarmPin
import app.farmsy.android.core.LanguageStore
import app.farmsy.android.core.ShoppingItems
import app.farmsy.android.core.ShoppingPlanner
import app.farmsy.android.features.shopping.LockedSample
import app.farmsy.android.ui.theme.FarmsyColors
import app.farmsy.android.ui.theme.PrimaryButton
import app.farmsy.android.ui.theme.card
import app.farmsy.android.ui.theme.geist
import com.google.android.gms.maps.model.LatLng
import kotlinx.coroutines.launch
import java.util.Locale

/// The planner's own radius — `ShoppingPlanner.plan`'s default, spelled here
/// because the coverage sentence has to name it.
private const val RADIUS_KM = 25.0

/// "I need eggs, milk and potatoes — where do I drive?" — mirrors iOS
/// ShoppingListSheet.
///
/// The map answers where a farm is. This answers which farms to visit for a
/// list, and in what order. The result is ordinary trip stops, so the route
/// line, the totals, the corridor list and the Google Maps hand-off all pick it
/// up unchanged.
///
/// A picker rather than a text field. The words that mean a product are served
/// (GET /api/shopping/items) because guessing them locally is what sends
/// somebody asking for lamb to a beef farm.
@OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)
@Composable
fun ShoppingListSheet(origin: LatLng, onUnlock: () -> Unit, onDismiss: () -> Unit) {
    val farms = LocalFarms.current
    val trip = LocalTrip.current
    val session = LocalSession.current
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    val wanted by trip.wantedProducts.collectAsState()
    val pins by farms.pins.collectAsState()
    val catalogue by ShoppingItems.items.collectAsState()
    val loadFailed by ShoppingItems.loadFailed.collectAsState()
    // Collected, not read off the store, so buying Plus mid-session unlocks
    // this sheet instead of leaving it blurred (final review #2).
    val profile by session.profile.collectAsState()
    val hasFullAccess = profile?.hasFullAccess == true

    var plan by remember { mutableStateOf<ShoppingPlanner.Plan?>(null) }
    var isPlanning by remember { mutableStateOf(false) }

    // Dutch or English, the same rule the website applies; fr and de fall back
    // to English because the labels only exist in two.
    val language = remember {
        LanguageStore.current(context).code.ifEmpty { Locale.getDefault().language }
    }

    // The produce text is what every match is made against — without it the sheet
    // would honestly report that nothing is sold anywhere.
    LaunchedEffect(Unit) {
        farms.loadFlagsIfNeeded()
        ShoppingItems.loadIfNeeded()
    }

    /// Ids back to the words on the chips, so the answer is read in the same
    /// language it was asked in.
    fun labels(ids: List<String>): String =
        ids.joinToString(" · ") { id -> ShoppingItems.item(id)?.label(language) ?: id }

    fun buildPlan() {
        isPlanning = true
        scope.launch {
            farms.loadFlagsIfNeeded()
            ShoppingItems.loadIfNeeded()
            val picked = wanted.mapNotNull { ShoppingItems.item(it) }
            val candidates = pins.map {
                ShoppingPlanner.Candidate(it.osmId, LatLng(it.lat, it.lng), farms.produceFor(it.osmId) ?: "")
            }
            plan = ShoppingPlanner.plan(wanted = picked, farms = candidates, origin = origin, radiusKm = RADIUS_KM)
            isPlanning = false
        }
    }

    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = sheetState, containerColor = FarmsyColors.cream) {
        Column(
            Modifier.fillMaxWidth().verticalScroll(rememberScrollState())
                .padding(horizontal = 20.dp).padding(bottom = 28.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            // Header
            Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                Column(Modifier.weight(1f)) {
                    Text(stringResource(R.string.shopping_list), style = geist(18.sp, FontWeight.Bold), color = FarmsyColors.ink)
                    Text(stringResource(R.string.shopping_list_sub), style = geist(12.sp), color = FarmsyColors.inkMuted)
                }
                Icon(
                    Icons.Filled.Close, null, tint = FarmsyColors.inkMuted,
                    modifier = Modifier.size(22.dp).clickable { onDismiss() },
                )
            }

            if (catalogue.isEmpty()) {
                if (loadFailed) {
                    Text(
                        stringResource(R.string.shopping_list_catalogue_failed),
                        style = geist(14.sp), color = FarmsyColors.inkMuted,
                    )
                } else {
                    Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                        CircularProgressIndicator(color = FarmsyColors.farmGreen)
                    }
                }
            } else {
                // The picker
                FlowRow(
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    catalogue.forEach { item ->
                        val on = item.id in wanted
                        Row(
                            Modifier
                                .background(if (on) FarmsyColors.farmGreenMap else Color.White, RoundedCornerShape(20.dp))
                                .border(1.dp, if (on) Color.Transparent else FarmsyColors.hairline, RoundedCornerShape(20.dp))
                                .clickable { trip.toggleProduct(item.id); plan = null }
                                .padding(horizontal = 12.dp, vertical = 8.dp),
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(6.dp),
                        ) {
                            if (on) {
                                Icon(Icons.Filled.Check, null, tint = Color.White, modifier = Modifier.size(12.dp))
                            }
                            Text(
                                item.label(language),
                                style = geist(14.sp, FontWeight.Medium),
                                color = if (on) Color.White else FarmsyColors.ink,
                            )
                        }
                    }
                }

                // The answer. Which farms cover the list is the paid half — the
                // same answer the Shopping tab blurs — so a free visitor gets
                // the identical locked sample instead of the named farms
                // (final review #1).
                plan?.let { p ->
                    if (hasFullAccess || p.isEmpty) {
                        ResultCard(p, pins, ::labels)
                    } else {
                        LockedSample(
                            coverage = stringResource(
                                R.string.shopping_sample_coverage_arg,
                                p.coveredCount, wanted.size, p.picks.size, RADIUS_KM.toInt(),
                            ),
                            label = stringResource(R.string.see_which_farms),
                            onUnlock = onUnlock,
                        ) {
                            ResultCard(p, pins, ::labels)
                        }
                    }
                }

                // Action. Planning is free (it is what draws the sample);
                // adding the farms it found to the trip is the paid half, so
                // for a free visitor the button says what it does and opens Plus.
                val ready = plan?.isEmpty == false
                val locked = ready && !hasFullAccess
                if (isPlanning) {
                    Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                        CircularProgressIndicator(color = FarmsyColors.farmGreen)
                    }
                } else {
                    PrimaryButton(
                        when {
                            locked -> stringResource(R.string.see_which_farms)
                            ready -> stringResource(R.string.shopping_list_add_stops_arg, plan!!.picks.size)
                            else -> stringResource(R.string.shopping_list_plan)
                        },
                        enabled = wanted.isNotEmpty(),
                    ) {
                        when {
                            locked -> onUnlock()
                            ready -> {
                                trip.addStops(plan!!.picks.map { it.osmId })
                                trip.requestFit()
                                onDismiss()
                            }
                            else -> buildPlan()
                        }
                    }
                }

                if (wanted.isNotEmpty()) {
                    Text(
                        stringResource(R.string.shopping_list_clear),
                        style = geist(13.sp, FontWeight.Medium), color = FarmsyColors.inkMuted,
                        modifier = Modifier.align(Alignment.CenterHorizontally).clickable {
                            trip.clearProducts(); plan = null
                        },
                    )
                }
            }
            Spacer(Modifier.height(4.dp))
        }
    }
}

/// The numbered farms the plan picked, and what it could not find. Drawn sharp
/// for a member and blurred inside `LockedSample` for everybody else — the same
/// rows either way, because the sample has to be the real answer.
@Composable
private fun ResultCard(
    plan: ShoppingPlanner.Plan,
    pins: List<FarmPin>,
    labels: (List<String>) -> String,
) {
    Column(
        Modifier.fillMaxWidth().card(16),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        if (plan.isEmpty) {
            Text(stringResource(R.string.shopping_list_none), style = geist(14.sp), color = FarmsyColors.inkMuted)
        } else {
            plan.picks.forEachIndexed { i, pick ->
                Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                    Box(
                        Modifier.size(24.dp).background(FarmsyColors.farmGreenMap, CircleShape),
                        contentAlignment = Alignment.Center,
                    ) {
                        Text("${i + 1}", style = geist(13.sp, FontWeight.Bold), color = Color.White)
                    }
                    Column(Modifier.weight(1f)) {
                        Text(
                            pins.firstOrNull { it.osmId == pick.osmId }?.name ?: pick.osmId,
                            style = geist(15.sp, FontWeight.SemiBold), color = FarmsyColors.ink, maxLines = 1,
                        )
                        Text(
                            labels(pick.covers),
                            style = geist(13.sp), color = FarmsyColors.inkMuted, maxLines = 2,
                        )
                    }
                }
            }
        }
        // Said out loud rather than quietly dropped: a list that half worked is
        // only useful if you know which half.
        if (plan.missing.isNotEmpty()) {
            Text(
                stringResource(R.string.shopping_list_missing_arg, labels(plan.missing)),
                style = geist(13.sp, FontWeight.Medium), color = FarmsyColors.inkMuted,
            )
        }
    }
}
